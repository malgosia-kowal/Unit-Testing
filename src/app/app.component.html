<div class="header">
  <h1>
    {{ title }}
  </h1>
</div>
<div class="container">
  <app-products></app-products>
  <app-quickview></app-quickview>
  <app-basket></app-basket>
</div>