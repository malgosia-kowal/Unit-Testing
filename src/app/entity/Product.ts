export class Product {
  public name: string;
  public price: number;
  public quantity: number;
  public size: string;
  public image: string;

  constructor(name: string, price: number, quantity: number, size: string, image: string) {
    this.name = name;
    this.price = price;
    this.quantity = quantity;
    this.size = size;
    this.image = image; 
  }
}
