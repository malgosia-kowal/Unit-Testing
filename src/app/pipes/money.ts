import { Pipe, PipeTransform } from "@angular/core";

@Pipe({ name: "money" })
export class MoneyPipe implements PipeTransform {
  transform(value: number): string {
    return (value / 100).toFixed(2).replace(".", ",");
  }
}
