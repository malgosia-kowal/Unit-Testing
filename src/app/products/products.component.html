<h2>Products</h2>
<app-product-detail [product]="product"></app-product-detail>
<div class="productsContainer">
  <div class="productCard" *ngFor="let product of products" (click)="onSelect(product)">
    <img src="{{ product.image }}">
    <h3>{{ product.name }}</h3>
    <hr>
    <div>
      <span class="productCardPropHeader">Color: </span>
      <span class="productCardPropValue">{{ product.size }}</span>
    </div>
    <div>
      <span class="productCardPropHeader">Price </span>
      <span class="productCardPropValue">{{ product.price }}</span> zl
    </div>
    <hr>
    <div class="productCardFooter">
      <app-button text="Add to basket" (action)="addProduct(product); $event.stopPropagation()">
      </app-button>
    </div>
  </div>
</div>