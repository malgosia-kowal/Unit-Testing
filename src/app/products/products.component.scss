.productsContainer {
  display: flex;
  flex-wrap: wrap;
  margin-bottom: 70px;
}

.productCard {
  padding: 10px;
  margin: 10px;
  box-shadow: 0 5px 10px rgb(0, 0, 0, 0.2);
  cursor: pointer;
  &:hover {
    box-shadow: 0 10px 15px rgb(0, 0, 0, 0.4);
  }

  &Footer {
    display: flex;
    flex-direction: row;
    justify-content: flex-end;
    & > div {
      padding: 5px;
      background: darkred;
      color: white;
    }
  }
}
