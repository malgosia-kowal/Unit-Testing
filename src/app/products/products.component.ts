import { Component, OnInit } from '@angular/core';
import { Product } from './product';
import { ProductService } from '../service/product.service';
import { BasketService } from '../service/basket.service';

@Component({
  selector: 'app-products',
  templateUrl: './products.component.html',
  styleUrls: ['./products.component.scss']
})
export class ProductsComponent implements OnInit {
  product: Product;
  products: Product[];

  constructor(private productService: ProductService, private basketService: BasketService) {
    this.productService = productService;
  }

  ngOnInit() {
    this.products = this.productService.getProducts();
  }

  onSelect(product: Product) {
    this.product = product;
  }

  addProduct(product: Product) {
    this.basketService.addProduct(product);
  }
}
