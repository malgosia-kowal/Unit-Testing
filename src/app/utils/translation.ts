import { TranslateService } from "@ngx-translate/core";
import { Locale } from "../app.component";

export const getPreviousLang = (
  translate: TranslateService,
  lang: string
): Locale => {
  return (translate.getLangs().filter(l => l !== lang)[0] ||
    translate.getDefaultLang()) as Locale;
};
