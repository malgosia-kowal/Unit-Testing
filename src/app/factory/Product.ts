import { Product } from '../entity/Product'; 
import * as faker from 'faker';

export const createProduct = (data: Partial<Product> = {}) => new Product(
    data.hasOwnProperty('name') ? data.name : faker.commerce.productName(), 
    data.hasOwnProperty('price') ? data.price : parseInt(faker.commerce.price()), 
    1, 
    data.hasOwnProperty('size') ? data.size : faker.commerce.color(),
    data.hasOwnProperty('image') ? data.image : faker.image.imageUrl(),
);
