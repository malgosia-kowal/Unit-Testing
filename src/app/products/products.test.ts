import { async, ComponentFixture, TestBed } from "@angular/core/testing";
import { ProductsComponent } from "./products.component";
import { ProductDetailComponent } from "../product-detail/product-detail.component";
import { BasketService } from "../service/basket.service";
import { ButtonComponent } from "../button/button.component";
import { ProductService } from "../service/product.service";
import { createProduct } from "../factory/Product";
import { By } from "@angular/platform-browser";
import {
  TranslateModule,
  TranslateLoader,
  TranslateService
} from "@ngx-translate/core";
import { MockCustomLoader } from "../tests/mocks/translation";
import { MoneyPipe } from "../pipes/money";
import { BehaviorSubject } from "rxjs";
import { Locale } from "../app.component";

class MockProductsService {
  getProducts = jest.fn(() => new BehaviorSubject([]));
  convertProductsPrice = jest.fn();
}

describe("Products Component", () => {
  let component: ProductsComponent;
  let fixture: ComponentFixture<ProductsComponent>;
  let basketService: BasketService;
  let productService: MockProductsService;
  let translate: TranslateService;

  beforeEach(async(() => {
    jest.resetAllMocks();
    TestBed.configureTestingModule({
      declarations: [
        ProductsComponent,
        ProductDetailComponent,
        ButtonComponent,
        MoneyPipe
      ],
      imports: [
        TranslateModule.forRoot({
          loader: { provide: TranslateLoader, useClass: MockCustomLoader }
        })
      ],
      providers: [
        { provide: ProductService, useClass: MockProductsService },
        { provide: BasketService, useClass: BasketService }
      ]
    })
      .compileComponents()
      .then(() => {
        fixture = TestBed.createComponent(ProductsComponent);
        component = fixture.componentInstance;

        basketService = TestBed.get(BasketService);
        productService = TestBed.get(ProductService);
        translate = TestBed.get(TranslateService);
        translate.use(Locale.Gb);

        fixture.detectChanges();
      });
  }));

  it("should exist", () => {
    expect(component).toBeDefined();
  });

  it("should fetch products on init", () => {
    expect(productService.getProducts).toHaveBeenCalledTimes(1);
  });

  it("should handle with no products", () => {
    expect(component.products).toHaveLength(0);

    expect(fixture).toMatchSnapshot();
  });

  it("can render products", () => {
    const product1 = createProduct({
      name: "test",
      size: "8",
      image: "url",
      price: 8
    });
    const product2 = createProduct({
      name: "test2",
      size: "8",
      image: "url",
      price: 4
    });
    productService.getProducts.mockImplementation(
      () => new BehaviorSubject([product1, product2])
    );

    component.ngOnInit();

    expect(component.products).toHaveLength(2);

    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it("add products to the basket", () => {
    const product1 = createProduct({
      name: "test2",
      size: "8",
      image: "url",
      price: 4
    });
    productService.getProducts.mockImplementation(
      () => new BehaviorSubject([product1])
    );

    component.ngOnInit();

    fixture.detectChanges();

    basketService.addProduct(product1);
    fixture.detectChanges();

    expect(basketService.get().products.getValue().length).toEqual(1);
  });

  it("can select product", () => {
    const product1 = createProduct({
      name: "test2",
      size: "8",
      image: "url",
      price: 4
    });
    const product2 = createProduct({
      name: "test3",
      size: "8",
      image: "url",
      price: 9
    });
    productService.getProducts.mockImplementation(
      () => new BehaviorSubject([product1, product2])
    );

    component.ngOnInit();

    fixture.detectChanges();

    expect(component.product).toBeUndefined();

    component.onSelect(product1);

    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();

    const ProductDetailComponent = fixture.debugElement.query(
      By.css("app-product-detail")
    ).componentInstance;

    expect(component.product).toEqual(product1);
    expect(ProductDetailComponent).toMatchObject({ product: product1 });
  });
});
