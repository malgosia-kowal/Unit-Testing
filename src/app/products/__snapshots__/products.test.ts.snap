// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Products Component can render products 1`] = `
<app-products
  basketService={[Function BasketService]}
  productService={[Function MockProductsService]}
  products={[Function Array]}
>
  <h2>
    Products
  </h2><app-product-detail>
    
  </app-product-detail><div
    class="productsContainer"
  >
    
    <div
      class="productCard"
    >
      <img
        src="url"
      />
      <h3>
        test
      </h3>
      <hr />
      <div>
        <span
          class="productCardPropHeader"
        >
          Color: 
        </span>
        <span
          class="productCardPropValue"
        >
          8
        </span>
      </div>
      <div>
        <span
          class="productCardPropHeader"
        >
          Price 
        </span>
        <span
          class="productCardPropValue"
        >
          8
        </span>
         zl 
      </div>
      <hr />
      <div
        class="productCardFooter"
      >
        <app-button
          id="addToBasketButton-0"
          ng-reflect-text="Add to basket"
          text="Add to basket"
        >
          <button>
             Add to basket
  
          </button>
        </app-button>
      </div>
    </div>
    <div
      class="productCard"
    >
      <img
        src="url"
      />
      <h3>
        test2
      </h3>
      <hr />
      <div>
        <span
          class="productCardPropHeader"
        >
          Color: 
        </span>
        <span
          class="productCardPropValue"
        >
          8
        </span>
      </div>
      <div>
        <span
          class="productCardPropHeader"
        >
          Price 
        </span>
        <span
          class="productCardPropValue"
        >
          4
        </span>
         zl 
      </div>
      <hr />
      <div
        class="productCardFooter"
      >
        <app-button
          id="addToBasketButton-1"
          ng-reflect-text="Add to basket"
          text="Add to basket"
        >
          <button>
             Add to basket
  
          </button>
        </app-button>
      </div>
    </div>
  </div>
</app-products>
`;

exports[`Products Component can select product 1`] = `
<app-products
  basketService={[Function BasketService]}
  product={[Function Product]}
  productService={[Function MockProductsService]}
  products={[Function Array]}
>
  <h2>
    Products
  </h2><app-product-detail
    ng-reflect-product="[object Object]"
  >
    
    <div>
       test2
  
    </div>
  </app-product-detail><div
    class="productsContainer"
  >
    
    <div
      class="productCard"
    >
      <img
        src="url"
      />
      <h3>
        test2
      </h3>
      <hr />
      <div>
        <span
          class="productCardPropHeader"
        >
          Color: 
        </span>
        <span
          class="productCardPropValue"
        >
          8
        </span>
      </div>
      <div>
        <span
          class="productCardPropHeader"
        >
          Price 
        </span>
        <span
          class="productCardPropValue"
        >
          4
        </span>
         zl 
      </div>
      <hr />
      <div
        class="productCardFooter"
      >
        <app-button
          id="addToBasketButton-0"
          ng-reflect-text="Add to basket"
          text="Add to basket"
        >
          <button>
             Add to basket
  
          </button>
        </app-button>
      </div>
    </div>
    <div
      class="productCard"
    >
      <img
        src="url"
      />
      <h3>
        test3
      </h3>
      <hr />
      <div>
        <span
          class="productCardPropHeader"
        >
          Color: 
        </span>
        <span
          class="productCardPropValue"
        >
          8
        </span>
      </div>
      <div>
        <span
          class="productCardPropHeader"
        >
          Price 
        </span>
        <span
          class="productCardPropValue"
        >
          9
        </span>
         zl 
      </div>
      <hr />
      <div
        class="productCardFooter"
      >
        <app-button
          id="addToBasketButton-1"
          ng-reflect-text="Add to basket"
          text="Add to basket"
        >
          <button>
             Add to basket
  
          </button>
        </app-button>
      </div>
    </div>
  </div>
</app-products>
`;

exports[`Products Component should handle with no products 1`] = `
<app-products
  basketService={[Function BasketService]}
  productService={[Function MockProductsService]}
  products={[Function Array]}
>
  <h2>
    Products
  </h2><app-product-detail>
    
  </app-product-detail><div
    class="productsContainer"
  >
    
  </div>
</app-products>
`;
